File: tests/test_misc.py
#!/usr/bin/env python

from helpers import *


class Test_Misc(BasisTestCase):
  def test_abc(self):
    np.random.seed(0)
    d1 = 3
    nki1 = 8
    k1 = np.r_[np.zeros(d1),np.linspace(0., 1., nki1), np.ones(d1)]
    b1 = BSplineBasis(k1, d1)
    C = casadi.MX(np.random.rand(b1.dimension()))
    f = Function(b1,C)
        
    f **2

  def test_jac_test(self):
    degree = 1
    knots = np.r_[np.zeros(degree), np.linspace(0, 1,2), np.ones(degree)]
    basis = BSplineBasis(knots, degree)

    X = MX.sym('X', basis.dimension())
    x = Function(basis, X)

    expr = x.derivative()
    expr = expr.data()

    jac = jacobian(expr, X)

    self.assertTrue(len(symvar(jac))==0)

    x = Function(basis, X**2)

    expr = x.derivative()
    expr = expr.data()

    jac = jacobian(expr, X)

    self.assertFalse(len(symvar(jac))==0)

  def test_numpy(self):
    np.random.seed(0)
    d1 = 3
    nki1 = 8
    k1 = np.r_[np.zeros(d1),np.linspace(0., 1., nki1), np.ones(d1)]
    b1 = BSplineBasis(k1, d1)
    C = casadi.MX(np.random.rand(b1.dimension()))
    f = Function(b1,C)
    np.array([f,f])<0
    pass
         
if __name__ == '__main__':        
    unittest.main()

File: tests/test_misc.m
meco_binaries('cpp_splines','fill_in_the_branch_you_want')

import splines.*

try
  F = splines.Function(3);
except
  
end

knots = linspace(0,1,20);

degree = 3;

B2 = BSplineBasis(knots,degree);


123

B2.knots()

b1 = BSplineBasis([0,0,1/2,1,1],1);
c1 = Coefficient(rand(3,1,1));
s1 = splines.Function(b1, c1);

coeff_tensor(s1+casadi.SX.sym('x',2))
coeff_tensor(casadi.SX.sym('x',2)+s1)

w = DTensor(rand(2,2,2))
w+casadi.SX.sym('x')
casadi.SX.sym('x')+w

b1 = BSplineBasis([0,0,1/2,1,1],1);
c1 = Coefficient(rand(3,4,2));
s1 = splines.Function(b1, c1);

assert(norm(size(s1)-[4,2])==0)
assert(norm(size(s1,1)-[4])==0)
assert(norm(size(s1,2)-[2])==0)
%assert(norm(numel(s1)-[8])==0)
assert(norm(size(s1(1:end,1:end),1)-[4])==0)
assert(norm(size(s1(1:end,1:end),2)-[2])==0)
assert(norm(size(s1(2:end,2:end),3)-[1])==0)
assert(norm(size(s1(2:end,2:end),2)-[1])==0)

a = DTensor(rand(4,4,2));
a(:)
